File: cpp_file/XIRHXQ_CVT.cpp
// Created by XIRHXQ on 12/29/21.

#include "XIRHXQ_CVT.h"

#include <functional>

XIRHXQ_CVT::XIRHXQ_CVT() {

}

XIRHXQ_CVT::XIRHXQ_CVT(int _n, const XIRHXQ_Polygon &_wd) {
    n = _n;
    wd = _wd;
    for (int i = 1; i <= n; i++){
        pt[i] = XIRHXQ_Point();
        pl[i] = _wd;
    }
}

XIRHXQ_CVT::XIRHXQ_CVT(int _n, const XIRHXQ_Point *_pt, const XIRHXQ_Polygon &_wd) {
    n = _n;
    wd = _wd;
    for (int i = 1; i <= n; i++){
        pt[i] = _pt[i - 1];
        pl[i] = _wd;
    }
}

void XIRHXQ_CVT::reset_point(int _n, const XIRHXQ_Point *_pt) {
    n = _n;
    for (int i = 1; i <= n; i++){
        pt[i] = _pt[i - 1];
        pl[i] = wd;
    }
}

void XIRHXQ_CVT::cal_poly() {
    XIRHXQ_Line l;
    for (int i = 1; i <= n; i++){
        for (int j = 1; j <= n; j++){
            if (i == j) continue;
            l = median_line(pt[i], pt[j]);
            pl[i].intersect_with_halfplane(l, pt[i]);
        }
    }
}

void XIRHXQ_CVT::output() {
    printf("An XIRHXQ_CVT with %d points:----------\n", n);
    for (int i = 1; i <= n; i++) {
        printf("Point #%d: ", i);
        pt[i].output();
        printf("Poly #%d: ", i);
        pl[i].output();
        printf("Centroid #%d: ", i);
        ct[i].output();
        printf("-------\n");
    }
    printf("---------------\n");
}

void XIRHXQ_CVT::cal_centroid(const std::function<double(XIRHXQ_Point)>& f, double spacing) {
    for (int i = 1; i <= n; i++){
        std::function<double(XIRHXQ_Point)> fx = [=](XIRHXQ_Point _p){return f(_p) * _p.x;};
        std::function<double(XIRHXQ_Point)> fy = [=](XIRHXQ_Point _p){return f(_p) * _p.y;};
        ct[i] = XIRHXQ_Point(pl[i].area_with_function(fx, spacing),
                             pl[i].area_with_function(fy, spacing)) /
                                     pl[i].area_with_function(f, spacing);
    }
}

File: XIRHXQ_CVT.m
classdef XIRHXQ_CVT < handle
    properties
        n(1, 1) {mustBeInteger}
        point XIRHXQ_Point
        poly XIRHXQ_Polygon
        world XIRHXQ_Polygon
    end
    methods
        function obj = XIRHXQ_CVT(varargin)
            if nargin == 2
                if isnumeric(varargin{1})
                    if size(varargin{1}, 2) ~= 2
                        error('Wrong Size of Input');
                    end
                    obj.n = size(varargin{1}, 1);
                    for i = obj.n:-1:1
                        obj.point(i) = XIRHXQ_Point(varargin{1}(i, :));
                    end
                else
                    obj.n = numel(varargin{1});
                    for i = obj.n:-1:1
                        obj.point(i) = varargin{1}(i);
                    end
                end
                obj.world = varargin{2}.copy;
                for i = obj.n:-1:1
                    obj.poly(i) = varargin{2}.copy;
                end
            else
                error('Invalid input');
            end
        end
        function obj = reset_point(obj, pts)
            if isnumeric(pts)
                for i = 1:obj.n
                    obj.point(i) = XIRHXQ_Point(pts(i, :));
                end
            else
                obj.point = pts;
            end
            for i = 1:obj.n
                obj.poly(i) = obj.world.copy;
            end
        end
        function obj = cal_poly(obj)
            for i = 1:obj.n
                for j = 1:obj.n
%                     fprintf('i = %d, j = %d\n', i, j);
                    if i == j
                        continue;
                    end
%                     obj.point(i)
%                     obj.point(j)
%                     XIRHXQ_Line.midean_line(obj.point(i), obj.point(j))
                    obj.poly(i).intersect_with_halfplane(XIRHXQ_Line.midean_line(obj.point(i), obj.point(j)), obj.point(i));
%                     obj.poly(i)
                end
            end
        end
        function disp(obj)
            fprintf('A XIRHXQ_CVT with %d points\n', obj.n);
            for i = 1:obj.n
                fprintf('Point #%d @ (%.2f, %.2f):\n', i, obj.point(i).x, obj.point(i).y);
                disp(obj.poly(i));
            end
        end
        function plot(obj)
%             figure('Visible', 'off');
            figure;
            for i = 1:obj.n
                plot(obj.poly(i));
                hold on;
            end
            [xlimit, ylimit] = obj.world.get_screen;
            xlim(xlimit);ylim(ylimit);
        end
        function c = get_centroid(obj, f)
            for i = obj.n:-1:1
                c(i,:) = XIRHXQ_Point([obj.poly(i).area_with_function(@(x)f(x)*x.x) obj.poly(i).area_with_function(@(x)f(x)*x.y)])...
                         / obj.poly(i).area_with_function(f);
            end
        end
    end
end